controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java

```diff
@@ -76,6 +80,106 @@
     private final static String RESOURCE_PART_SEPARATOR = "_%_";
 
+    @Data
+    /**
+     * This class represents a 96 bit number with 32 bit msb encoded as integer and 64 bit lsb encoded as long.
+     * It is unsigned and only allows for non negative values. It also implements comparable interface and comparison involves
+     * first compariging msbs and if msbs are equal then we compare lsbs.
+     */
+    protected static class BigLong implements Comparable {
+        public static final BigLong ZERO = new BigLong(0, 0L);
+        private final int msb;
+        private final long lsb;
+
+        BigLong(int msb, long lsb) {
+            Preconditions.checkArgument(msb >= 0);
+            Preconditions.checkArgument(lsb >= 0);
+
+            this.msb = msb;
+            this.lsb = lsb;
+        }
+
+        @Override
+        public int compareTo(Object o) {
+            if (!(o instanceof BigLong)) {
+                throw new RuntimeException("incomparable objects");
+            }
+            BigLong other = (BigLong) o;
+
+            if (msb != other.msb) {
+                return Integer.compare(msb, other.msb);
+            } else {
+                return Long.compare(lsb, other.lsb);
+            }
+        }
+
+        byte[] toBytes() {
+            byte[] b = new byte[Integer.BYTES + Long.BYTES];
+            BitConverter.writeInt(b, 0, msb);
+            BitConverter.writeLong(b, Integer.BYTES, lsb);
+            return b;
+        }
+
+        static BigLong fromBytes(byte[] b) {
+            int msb = BitConverter.readInt(b, 0);
+            long lsb = BitConverter.readLong(b, Integer.BYTES);
+
+            return new BigLong(msb, lsb);
+        }
+
+        static BigLong add(BigLong counter, int increment) {
+            BigLong retVal;
+            if (counter.lsb < Long.MAX_VALUE - increment) {
+                retVal = new BigLong(counter.msb, counter.lsb + increment);
+            } else if (counter.msb < Integer.MAX_VALUE) {
+                int remainder = increment - (int) (Long.MAX_VALUE - counter.lsb);
+                retVal = new BigLong(counter.msb + 1, remainder);
+            } else {
+                // overflow: throw exception
+                throw new ArithmeticException("Overflow");
+            }
+
+            return retVal;
+        }
+    }
+
+    /**
+     * This class provides the ability to atomically update a BigLong value.
+     */
+    protected static class AtomicBigLong {
+        @GuardedBy("lock")
+        private BigLong value;
+        private final Object lock = new Object();
+
+        AtomicBigLong() {
+            this.value = BigLong.ZERO;
+        }
+
+        @VisibleForTesting
+        AtomicBigLong(int msb, long lsb) {
+            Preconditions.checkArgument(msb >= 0);
+            Preconditions.checkArgument(lsb >= 0);
+            this.value = new BigLong(msb, lsb);
+        }
+
+        BigLong get() {
+            return this.value;
```

Comment:
This must be under the lock. Or just just AtomicReference.

Reply:
the value we return is an immutable object. So get can get a stale value if a concurrent update is taking place. And we are ok with that. hence no locking here.

Reply:
value may be immutable, but not its pointer. If the field itself would be marked `final`, this would be fine. But since it's not, you need to access it under guard.